### ISSUE TITLE
Pasting from Windows Terminal removes newlines

### ISSUE THREAD

**Issue by 89z** [microsoft/terminal#9933]
Using this program:

https://github.com/vim/vim-win32-installer/releases/download/v8.2.2800/gvim_8.2.2800_x64.zip

If I extract and start `vim.exe` inside Windows Terminal, I see this:

~~~
    version 8.2.2800
by Bram Moolenaar et al.
~~~

I have `copyOnSelect` enabled, so I select this text, then paste into Notepad. Result:

~~~
version 8.2.2800                                                    ~                                               by Bram Moolenaar et al.
~~~

Notice everything is on a single line. I tried disabling `copyOnSelect` and using Ctrl+C, but same result. I saw someone say that you have to hold Shift, I tried that and it doesnt work either. I also saw someone mention `singleLine`, but my setting seems to be correct:

~~~json
{ "command": {"action": "copy", "singleLine": false }, "keys": "ctrl+c" },
~~~

**Comment by 89z** [comment]
https://github.com/noborus/ov/issues/81

**Comment by 89z** [comment]
https://github.com/gdamore/tcell/issues/451

**Comment by zadjii-msft** [comment]
Wait how _specifically_ are you copying? With right click, with <kbd>ctrl+c</kbd>, with <kbd>ctrl+shift+c</kbd>, with copy on select? Could you share your `settings.json`?

**Comment by zadjii-msft** [comment]
(this has flavors of #9696 in it)

**Comment by 89z** [comment]
> Wait how _specifically_ are you copying? With right click, with ctrl+c, with ctrl+shift+c, with copy on select?

I tried copy on select, and Ctrl+C

> Could you share your `settings.json`?

~~~json
{
   "$schema": "https://aka.ms/terminal-profiles-schema",
   "copyOnSelect": true,
   "multiLinePasteWarning": false,
   "profiles": {
      "defaults": {},
      "list": [
         {
            "commandline": "C:/PowerShell/pwsh.exe -NoLogo",
            "fontSize": 11,
            "guid": "{61c54bbd-c2c6-5271-96e7-009a87ff44bf}"
         }, {
            "guid": "{b453ae62-4e3d-5e58-b989-0a998ec441b8}",
            "hidden": false,
            "name": "Azure Cloud Shell",
            "source": "Windows.Terminal.Azure"
         }
      ]
   },
   "keybindings": [
      { "command": {"action": "copy", "singleLine": false }, "keys": "ctrl+c" }
   ]
}
~~~

**Comment by 89z** [comment]
https://github.com/walles/moar/issues/51

**Comment by 89z** [comment]
Looks like this has been a problem for a while, I tried these versions, and they all have the same problem:

~~~
v1.8.1444.0
v1.4.3243.0
v1.3.2382.0
v1.2.2022.0
v1.1.1671.0
v1.0.1401.0
~~~

**Comment by 89z** [comment]
https://github.com/microsoft/terminal/issues/174

**Comment by 89z** [comment]
Closing for lack of interest

**Comment by zadjii-msft** [comment]
note to self: 
* still repros in 1.16 
* repros in conhost with quick edit manually enabled
* I vaguely remember an issue in the past where vim.exe does not use newlines to actually break rows. IIRC, they know that the line is about to wrap, so they let it just happen. 
* repros with `vim` in WSL too

**Comment by lhecker** [comment]
I don't believe this can be considered a bug in Windows Terminal as it happens in all other terminals as well without further integration into or configuration of vim. Searching for this issue on the internet brings up a lot of helpful advice how to fix this issue. One in particular that I liked is: https://stackoverflow.com/a/4608206
After running `:vmap <C-C> "+y`, I can simply select some text with my cursor (this enters the visual mode) and press Ctrl+C to copy to my clipboard. This works for me even if `copyOnSelect` is enabled, but one thing is a bit confusing: vim doesn't color the cursor position, so if you use the default I-beam (the thin vertical "|" line) it seems like one less character is selected, but as far as I can see that's exactly what `vim` wants us to draw.
Would you say that we should close this issue @89z? We can leave it open, but I don't believe we can do anything about this on our side at this time.

**Comment by 89z** [comment]
You seem to be getting hung up on the the Vim part of this issue, which is a classic mistake in troubleshooting.

Further, you seem to come to the conclusion that "well if all terminals are broken, then no one needs to fix anything!" which hopefully can be recognized as a silly conclusion. The issue isn't specific to Vim, thats just a good example because its a popular tool that demonstrates the problem. The issue is that any tool which activates "interactive mode", will suffer from this problem.

**Comment by 89z** [comment]
You seem to be getting hung up on the the Vim part of this issue, which is a classic mistake in troubleshooting.

Further, you seem to come to the conclusion that "well if all terminals are broken, then no one needs to fix anything!" which hopefully can be recognized as a silly conclusion. The issue isn't specific to Vim, thats just a good example because its a popular tool that demonstrates the problem. The issue is that any tool which activates "interactive mode", will suffer from this problem.

I tried to make it clear that this wasn't a specific issue with Vim, but then @zadjii-msft went and changed the title, which has only helped with the confusion.

**Comment by zadjii-msft** [comment]
At no point did you mention that this occurs with other applications? Your repro steps specifically called out vim, without mentioning anything else that repros this.

Vim does play silly games with line wrapping, hence my assumption of the root cause here. Other applications that play the same silly games would also experience the same issues, but without knowing where else you're seeing this, all we an do is presume that this is an issue specific to vim.

Do you have another commandline tool that reproduces the same behavior?

**Comment by 89z** [comment]
MITM Proxy demonstrates the same problem. Thats actually the tool that spurned this issue in the first place. I only changed it to Vim, because thats more "known". @zadjii-msft perhaps you could have just asked me before changing the title?

Like I said, its a classic mistake in troubleshooting. people have a tendency to want to blame something, in this case Vim, rather than consider a broader issue, or the simple fact that Vim might be acting "normally" and that its the Terminal doing something wrong. Because if Vim is the culprit, you can just close the issue and be done with it. But if its actually the Terminal, then work needs to be done. I'm not blaming anyone, but its something to keep in mind in the future.

**Comment by lhecker** [comment]
Could you expand on how we could solve this issue in technical terms given only the VT output / console API calls of `vim` without modifying `vim` itself (or "MITM proxy")?

**Comment by 89z** [comment]
@lhecker I don't really care with how to entered the conversation, so I not inclined to speak with you at all. However in the interest of at least saying "I tried", I will try. However I want to make sure we're on the same page before I put forth any effort. From your comments, you seem to want me to solve the issue for you, which I am also not inclined to do. *You are a member, not me*. As far as I'm concerned, my duty stops and ends with posting an issue. However I also understand that you might be inclined to close any issues where the poster doesn't solve the issue for you.

Is that your situation, or are you willing to let others work on this without me giving a solution. Put another way, do you require all issue posters to solve their own issue? Again I am will to put work in service of a goal here, but I don't like the way you entered the conversation, so I want to make damn sure I'm not wasting my time before making the decision to do that work.

**Comment by DHowett** [comment]
Hey @89z! I'm sorry. We didn't mean to offend you.

There's a constellation of issues here, all of which present in the same way. All of them involve applications that write to the entire screen (using Windows console APIs or by switching to the alternate screen buffer mode) and cause a set or reset of the "hard wrap" flag when they touch the end of the line. By and large, the decision we make here in the absence of line ending information is informed by what would be best for most applications.

We need to approach this from a few angles (simultaneously):
1. Ensure that applications are using the best available mechanism _today_ for reporting line wrapping, through community outreach and targeted work with those projects.
2. Better heuristics ("when `xyz.exe` touches the end of the buffer to draw spaces using `WriteConsoleXyz` API, know to ignore the wrap flag")
3. Better diagnostics (like: how did we not know that mitmproxy was in the same boat? Could we have caught it earlier?)

To Leonard's question: the burden is of course not on you, but if you do have any ideas as to how we can "know better" we would love to hear them. Not because it's your job, but because Terminal is a community and we want to make the best use of all our collective knowledge. :)

**Comment by 89z** [comment]
OK, here is one way to repro the issue. First, as I originally said, get Vim [1], and Alacritty [2]. Start Alacritty, then start Vim. Then enter copy mode with Alacritty (Ctrl+Shift+Space) and select the text mentioned in the original post. Copy with `y`. Then paste somewhere else. Result should be:

~~~
~                               version 8.2.2800
~                           by Bram Moolenaar et al.
~~~

Then start Windows Terminal, and start Vim. Select the text mentioned in the original post. I have `copyOnSelect`, so selection is copied. Then paste somewhere else. Result should be:

~~~
~                                version 8.2.2800                                  ~                            by Bram Moolenaar et al.
~~~

as mentioned in previous comments, this is just one example. Please do not use this comment as an excuse to target Alacritty, as I think other examples could probably be constructed with other terminals as well.

1. <https://github.com/vim/vim-win32-installer/releases/download/v8.2.2800/gvim_8.2.2800_x64.zip>
2. https://github.com/alacritty/alacritty/releases/download/v0.11.0/Alacritty-v0.11.0-portable.exe
